packages/modular-scripts/src/typecheck.ts

```diff
@@ -5,35 +5,107 @@ import getPackageMetadata from './utils/getPackageMetadata';
 import * as logger from './utils/logger';
 import getModularRoot from './utils/getModularRoot';
 import actionPreflightCheck from './utils/actionPreflightCheck';
+import { getAllWorkspaces } from './utils/getAllWorkspaces';
+import { selectWorkspaces } from './utils/selectWorkspaces';
 
-async function typecheck(): Promise<void> {
-  const { typescriptConfig } = await getPackageMetadata();
+import type { JSONSchemaForTheTypeScriptCompilerSConfigurationFile as TSConfig } from '@schemastore/tsconfig';
+
+type CompilerOptions = TSConfig['compilerOptions'];
 
-  const { _compilerOptions, ...rest } = typescriptConfig;
+export interface TypecheckOptions {
+  ancestors: boolean;
+  descendants: boolean;
+  changed: boolean;
+  compareBranch: string;
+}
 
-  const tsConfig: typeof typescriptConfig = {
+const COMPILER_OPTIONS_ALLOW_LIST: string[] = ['jsx'];
+const DEFAULT_COMPILER_OPTIONS: CompilerOptions = {
+  noEmit: true,
+};
+const DEFAULT_EXCLUDES: string[] = [
+  'node_modules',
+  'bower_components',
+  'jspm_packages',
+  'tmp',
+  '**/dist-types',
+  '**/dist-cjs',
+  '**/dist-es',
+  'dist',
+  '**/__fixtures__',
+];
+
+function restrictUserTsconfig(
+  userTsConfig: TSConfig,
+  replaceInclude: string[] | undefined,
+): TSConfig {
+  const { compilerOptions, ...rest } = userTsConfig;
+
+  // The exclude list here is different to that in getPackageInfo() because we want
+  // to test more files during a typecheck compared to a build (test/spec files). During a build,
+  // there might be various files that we don't want to end up in the output.
+  const tsConfig: TSConfig = {
     ...rest,
-    exclude: [
-      'node_modules',
-      'bower_components',
-      'jspm_packages',
-      'tmp',
-      '**/dist-types',
-      '**/dist-cjs',
-      '**/dist-es',
-      'dist',
-      '**/__fixtures__',
-    ],
-    compilerOptions: {
-      noEmit: true,
-    },
+    exclude: [...DEFAULT_EXCLUDES],
+  };
+
+  const combinedCompilerOptions: CompilerOptions = {
+    ...DEFAULT_COMPILER_OPTIONS,
   };
 
-  // If compilerOptions.jsx is defined use it otherwise, depend on the extended config by not setting it to undefined.
-  if (tsConfig.compilerOptions && rest.compilerOptions?.jsx) {
-    tsConfig.compilerOptions.jsx = rest.compilerOptions.jsx;
+  // Where the user has defined allowlist-supported compiler options, apply them
+  for (const item of COMPILER_OPTIONS_ALLOW_LIST) {
```

Comment:
This block is currently uncovered, worth adding a test case for it (essentially, using the jsx option)